README.md
bag.js - JS / CSS loader + KV storage
-------------------------------------

[![Build Status](https://travis-ci.org/nodeca/bag.js.svg?branch=master)](https://travis-ci.org/nodeca/bag.js)
![Version](https://badge.fury.io/gh/nodeca%2Fbag.js.svg)

__bag.js__ is loader for js / css aand other files, that uses browser local
stores for caching. Consider it as alternative for other types of loaders for
modern browsers, that reduce number of server requests, especially for mobile
devices. Also __bag.js__  can be used as simple key/value storage, that don't
require you to know details about IndexedDB and WebSQL.

This project is inspired by [basket.js](http://addyosmani.github.io/basket.js/),
but provides more storages for big assets and universal key/value interface.
Key features are:

- Parallel load and sequential execution for JS / CSS and other types of files
- Use IndexedDB / WebSQL / localStorage - good, when you have big assets.
- KV storage for objects, with simple interface.
- No promisses, use callbacks instead, like in node.js
- Simple way to prioritize scripts execution.
- You can use multiple instances with different storage options. For example
  Indexeddb + WebSQL for assets and localStorage for user settings.
- No external dependencies.
- 3.2K when minified+gzipped
- Partial compatibility with [basket.js](http://addyosmani.github.io/basket.js/).

Install via bower:

```
bower install bag.js
```


### Examples

Simple:

``` javascript
var bag = new window.Bag();

bag.require(['/site.css', '/jquery.js', '/site.js'], function (err) {
  if (err) {
    console.log('loading error: ', err);
    return
  }
  // code to run after loading
  // ...
})

```

Advanced:

``` javascript
var bag = new window.Bag({
  prefix: 'my_namespace',
  stores: ['indexeddb', 'websql'],
  timeout: 20000,
  expire: 24
});

bag.isValidItem = function(source, obj) {
  return (source && (source.url === obj.url)) ? true : false;
};

var files = [
  { url: '/site.css', expire: 1 },
  { url: '/jquery.js', expire: 10 },
  { url: '/site.js' },
  { url: '/more_styles.css', expire: 5, execute: false }  
];

bag.require(files, function(err) {
  if (err) {
    console.log('loading error: ', err);
    return
  }
  // code to run after loading
  // ...
})
```

You can skip `new` keyword. Aslo, you can use chained style - when `require`
called without callback, info is just remembered in internal stack:

``` javascript

window.Bag()
  .require('/site.css')
  .require('/jquery.js')
  .require('/site.js')
  .require(function (err, data) {
    if (err) { return console.log('loading error: ', err); }
    // code to run after loading
    // ...
  });
```

Using as key/value storage:

``` javascript
var obj = { lorem: 'ipsum' };
var bag = new window.Bag();

bag.set('dolorem', obj, function(err) {
  if (err) {
    console.log('Saving error: ', err);
    return;
  }

  bag.get('dolorem', function(err, data) {
    if (err) {
      console.log('Loading error: ', err);
      return;
    }

    console.log('Loaded data:\n', data);

    bag.remove('dolorem', function(err) {
      if (err) {
        console.log('Removing error: ', err);
        return;
      }

      console.log('Compleete');
    }
  }
});
```

API
---

### new Bag([options])

Object constructor. You can also define options after constructor call, via
instance properties (they have the same names). Options (hash):

- `prefix` - Data namespace. Default - `bag`. Used to separate data for
   multiple instances.
- `stores` - Array of storage names to use, ordered by preference.
  Default `['indexeddb', 'websql', 'localstorage']`.
- `timeout` - files loading timeout, in seconds. Default 20.
- `expire` - `require()` data expiration, in hours. Default - 1 month. 0 or
  unset - don't expire.

Note 1: you can skip `new` keyword, calling `Bag()` will return you new instance anyway.

Note 2: `prefix` must be set before `require`/`get`/`set`/`remove`/`clear` calls. Other options can be changed anytime.

### .require(files [, callback(err, data)])

1. Load files from server or from cache.
2. Inject known types into page (js/css by default), if execution not disabled.
   When multiple files requested (files are `Array`), those are loaded in
   parallel, but injected in defined order.
3. Return result in callback.

`files` param can be:

- `Object` - resource info (see details below).
- `String` - just resource url, other params will be default.
- `Array(Object|String)` - list of resources to load in parallel.

resource info:

- `url` - resource URI, required.
- `expire` - optional, expiration time in hours. 0 or not set - don't expire.
- `key` - the name, used to store loaded file, if not defined, then `url`
   will be used.
- `unique` - a token stored with the cached item. If you request the same item
  again with a different token the script will be fetched and cached again.
- `live` - force cache bypass, for development needs.
- `cached` - force request from cache only.

callback params:

- `err` - error info if loading failed
- `data` - (Array|String) with loaded content, dependong on `files` type. When
  single resource requester (`Object`|`String`), `data` is `String`. When
  `Array` of resourses requested, or chained call done, data is array of strings.

Note, if you pass resources info not in short form, input objects are extended
with loaded data.


### .get(key, callback(err, data))

Load data by `key` name. Return result via callback:

- `err` - error info on fail, null on success
- `data` - loaded object, `undefined` if not exists.


### .set(key, data [, expire] [, callback(err)])

Put data into storage under `key` name.

- `key` - String to address data.
- `data` - JS object to store. We currently support only objects, serializeable
  by JSON. Don't try to put functions or arraybuffers.
- `expire` - Expiration time in seconds. Don't expire by default.
- `callback(err)` - Function to call when compleete. `err` contains error
  on fail.


### .remove(key, callback(err))

Remove `key` data from store. Call `callback` when compleete. If error happens,
error info passed in first callback's param.


### .clear([expiredOnly] [, callback(err)])

Clear all storage data (in your namespace), or just expired objects when called
as `bag.clear(true, ...)`. Callback is optional.


### .addHandler(types, handler)

Add handler for loaded files with specified mime types. By default, handlers
for `application/javascript` and `text/css` already exist. If you set
`execute: false` in resource info, then handler will not be applied.

- `types` - `String` with mime type or `Array` of strings.
- `handler` - function to "execute" file of that type.


### .removeHandler(types)

Remove handler for specified mime type (opposite to `addHandler`).


Related projects
----------------

- [basket.js](http://addyosmani.github.io/basket.js/)
- [PortableCache.js](https://github.com/agektmr/PortableCache.js)
- [Lawnchair](http://brian.io/lawnchair/)
- [LABjs](https://github.com/getify/LABjs)
- [yepnope.js](https://github.com/SlexAxton/yepnope.js)


License
-------

[MIT](https://github.com/nodeca/bag.js/blob/master/LICENSE)


HISTORY.md
0.1.3 / 2014-12-12
------------------

- Added processing for outdated JS mime types (can happen on CDNs).
- Changed wrapper to be more universal (to support amd & browserify).


0.1.2 / 2014-03-25
------------------

- Care about embedded urls for soure maps (do address absolute on js/css ingect).


0.1.1 / 2013-11-30
------------------

- Fixed prefixes support.
- More tests.
- Cleanup.


0.1.0 / 2013-11-30
------------------

- First release.


test/test.js

describe('localStorage tests', function () {

  var bag = new window.Bag({
    stores: [ 'localstorage' ]
  });

  var key = 'test_key';
  var obj = { lorem: 'imsum', dolorem: 'casum' };
  var obj2 = { tratum: 'curem', lafem: 'pendum' };


  it('key set', function (done) {
    bag.set(key, obj, function (err) {
      assert.notOk(err);
      done();
    });
  });


  it('key get', function (done) {
    bag.get(key, function (err, data) {
      assert.notOk(err);
      assert.deepEqual(obj, data);
      done();
    });
  });


  it('key update', function (done) {
    bag.set(key, obj2, function (err) {
      assert.notOk(err);
      bag.get(key, function (err, data) {
        assert.notOk(err);
        assert.deepEqual(obj2, data);
        done();
      });
    });
  });


  it('key remove', function (done) {
    bag.remove(key, function (err) {
      assert.notOk(err);

      bag.get(key, function(err, data) {
        assert.ok(err);
        assert.notOk(data);
        done();
      });
    });
  });

});


test/test_extras.js
/*global describe, it, assert*/
'use strict';

describe('extra tests', function () {


  it('use localstorage by priority', function (done) {
    var b = new window.Bag({ stores: [ 'localstorage', 'websql' ] });
    b.set('ls_test', 'ls_test', function () {
      var b1 = new window.Bag({ stores: [ 'localstorage' ] });
      b1.get('ls_test', function (err, data) {
        assert.notOk(err);
        assert.strictEqual(data, 'ls_test');
        b.clear(done);
      });
    });
  });


  it('use websql by priority', function (done) {
    var b = new window.Bag({ stores: [ 'websql', 'localstorage' ] });
    b.set('wsql_test', 'wsql_test', function () {
      var b1 = new window.Bag({ stores: [ 'websql' ] });
      b1.get('wsql_test', function (err, data) {
        assert.notOk(err);
        assert.strictEqual(data, 'wsql_test');
        b.clear(done);
      });
    });
  });


  it('reset localstorage (namespace) when quota exceeded (2.5-5Mb)', function (done) {
    var b = new window.Bag({ stores: [ 'localstorage' ] });

    // create big data (>5mb in json)
    var huge = new Array(1000000);
    for (var i = 0, l = huge.length; i < l; i++) { huge[i] = '1234567890'; }

    b.set('permanent', 'permanent', function (err) {
      assert.notOk(err);
      b.set('tmp', 'tmp', 15, function (err) {
        assert.notOk(err);
        b.set('huge', huge, function (err) {
          assert.ok(err);

          b.get('tmp', function (err) {
            assert.ok(err);
            b.get('permanent', function (err) {
              assert.ok(err);
              b.clear(done);
            });
          });
        });
      });
    });
  });


  it('test namespace (only for localStorage)', function (done) {
    var b = new window.Bag({ prefix: 'ns', stores: [ 'localstorage' ] });

    b.set('key', 'value', function () {
      assert.ok(localStorage.ns__key);
      var val = JSON.parse(localStorage.ns__key).value;
      assert.strictEqual(val, 'value');
      b.clear(done);
    });
  });


  it('init without `new` keyword', function() {
    /*eslint-disable new-cap*/
    assert.instanceOf(new window.Bag(), window.Bag);
    assert.instanceOf(window.Bag(), window.Bag);
  });


});


bag.js
/*
 * bag.js - js/css/other loader + kv storage
 *
 * Copyright 2013-2014 Vitaly Puzrin
 * https://github.com/nodeca/bag.js
 *
 * License MIT
 */

/*global define*/

(function (root, factory) {
  'use strict';

  if (typeof define === 'function' && define.amd) {
    define(factory);
  } else if (typeof module === 'object' && typeof module.exports === 'object') {
    module.exports = factory();
  } else {
    root.Bag = factory();
  }
} (this, function () {
  'use strict';

  var head = document.head || document.getElementsByTagName('head')[0];

  //////////////////////////////////////////////////////////////////////////////
  // helpers

  function _nope() { return; }

  function _isString(obj) {
    return Object.prototype.toString.call(obj) === '[object String]';
  }

  var _isArray = Array.isArray || function isArray(obj) {
    return Object.prototype.toString.call(obj) === '[object Array]';
  };

  function _isFunction(obj) {
    return Object.prototype.toString.call(obj) === '[object Function]';
  }

  function _each(obj, iterator) {
    if (_isArray(obj)) {
      if (obj.forEach) {
        return obj.forEach(iterator);
      }
      for (var i = 0; i < obj.length; i++) {
        iterator(obj[i], i, obj);
      }
    } else {
      for (var k in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, k)) {
          iterator(obj[k], k);
        }
      }
    }
  }

  function _default(obj, src) {
    // extend obj with src properties if not exists;
    _each(src, function(val, key) {
      if (!obj[key]) { obj[key] = src[key]; }
    });
  }


  function _asyncEach(arr, iterator, callback) {
    callback = callback || _nope;
    if (!arr.length) { return callback(); }

    var completed = 0;
    _each(arr, function (x) {
      iterator(x, function (err) {
        if (err) {
          callback(err);
          callback = _nope;
        } else {
          completed += 1;
          if (completed >= arr.length) {
            callback();
          }
        }
      });
    });
  }


  //////////////////////////////////////////////////////////////////////////////
  // Adapters for Store class

  var DomStorage = function (namespace) {
    var self = this;
    var _ns = namespace + '__';
    var _storage = localStorage;


    this.init = function (callback) {
      callback();
    };


    this.remove = function (key, callback) {
      callback = callback || _nope;
      _storage.removeItem(_ns + key);
      callback();
    };


    this.set = function (key, value, expire, callback) {
      var obj = {
        value: value,
        expire: expire
      };

      var err;

      try {
        _storage.setItem(_ns + key, JSON.stringify(obj));
      } catch (e) {
        // On quota error try to reset storage & try again.
        // Just remove all keys, without conditions, no optimizations needed.
        if (e.name.toUpperCase().indexOf('QUOTA') >= 0) {
          try {
            _each(_storage, function(val, name) {
              var k = name.split(_ns)[1];
              if (k) { self.remove(k); }
            });
            _storage.setItem(_ns + key, JSON.stringify(obj));
          } catch (e2) {
            err = e2;
          }
        } else {
          err = e;
        }
      }

      callback(err);
    };


    this.get = function (key, raw, callback) {
      if (_isFunction(raw)) {
        callback = raw;
        raw = false;
      }

      var err, data;

      try {
        data = JSON.parse(_storage.getItem(_ns + key));
        data = raw ? data : data.value;
      } catch (e) {
        err = new Error('Can\'t read key: ' + key);
      }

      callback(err, data);
    };


    this.clear = function (expiredOnly, callback) {
      var now = +new Date();

      _each(_storage, function(val, name) {
        var key = name.split(_ns)[1];

        if (!key) { return; }

        if (!expiredOnly) {
          self.remove(key);
          return;
        }

        var raw;
        self.get(key, true, function(__, data) {
          raw = data; // can use this hack, because get is sync;
        });
        if (raw && (raw.expire > 0) && ((raw.expire - now) < 0)) {
          self.remove(key);
        }
      });

      callback();
    };
  };


  DomStorage.prototype.exists = function() {
    try {
      localStorage.setItem('__ls_test__', '__ls_test__');
      localStorage.removeItem('__ls_test__');
      return true;
    } catch (e) {
      return false;
    }
  };



  var WebSql = function (namespace) {
    var db;


    this.init = function (callback) {
      db = window.openDatabase(namespace, '1.0', 'bag.js db', 2e5);

      if (!db) { return callback('Can\'t open webdql database'); }

      db.transaction(function (tx) {
        tx.executeSql(
          'CREATE TABLE IF NOT EXISTS kv (key TEXT PRIMARY KEY, value TEXT, expire INTEGER KEY)',
          [],
          function () { return callback(); },
          function (tx, err) { return callback(err); }
        );
      });
    };


    this.remove = function (key, callback) {
      callback = callback || _nope;
      db.transaction(function (tx) {
        tx.executeSql(
          'DELETE FROM kv WHERE key = ?',
          [ key ],
          function () { return callback(); },
          function (tx, err) { return callback(err); }
        );
      });
    };


    this.set = function (key, value, expire, callback) {
      db.transaction(function (tx) {
        tx.executeSql(
          'INSERT OR REPLACE INTO kv (key, value, expire) VALUES (?, ?, ?)',
          [ key, JSON.stringify(value), expire ],
          function () { return callback(); },
          function (tx, err) { return callback(err); }
        );
      });
    };


    this.get = function (key, callback) {
      db.readTransaction(function (tx) {
        tx.executeSql(
          'SELECT value FROM kv WHERE key = ?',
          [ key ],
          function (tx, result) {
            if (result.rows.length === 0) {
              return callback(new Error('key not found: ' + key));
            }
            var value = result.rows.item(0).value;
            var err, data;
            try {
              data = JSON.parse(value);
            } catch (e) {
              err = new Error('Can\'t unserialise data: ' + value);
            }
            callback(err, data);
          },
          function (tx, err) { return callback(err); }
        );
      });
    };


    this.clear = function (expiredOnly, callback) {

      db.transaction(function (tx) {
        if (expiredOnly) {
          tx.executeSql(
            'DELETE FROM kv WHERE expire > 0 AND expire < ?',
            [ +new Date() ],
            function () { return callback(); },
            function (tx, err) { return callback(err); }
          );
        } else {
          db.transaction(function (tx) {
            tx.executeSql(
              'DELETE FROM kv',
              [],
              function () { return callback(); },
              function (tx, err) { return callback(err); }
            );
          });
        }
      });
    };
  };


  WebSql.prototype.exists = function() {
    return (!!window.openDatabase);
  };



  var Idb = function (namespace) {
    var db;

    this.init = function (callback) {
      var idb = this.idb = window.indexedDB; /* || window.webkitIndexedDB ||
                           window.mozIndexedDB || window.msIndexedDB;*/

      var req = idb.open(namespace, 2 /*version*/);

      req.onsuccess = function(e) {
        db = e.target.result;
        callback();
      };
      req.onblocked = function(e) {
        callback(new Error('IndexedDB blocked. ' + e.target.errorCode));
      };
      req.onerror = function(e) {
        callback(new Error('IndexedDB opening error. ' + e.target.errorCode));
      };
      req.onupgradeneeded = function(e) {
        db = e.target.result;
        if (db.objectStoreNames.contains('kv')) {
          db.deleteObjectStore('kv');
        }
        var store = db.createObjectStore('kv', { keyPath: 'key' });
        store.createIndex('expire', 'expire', { unique: false });
      };
    };


    this.remove = function (key, callback) {
      var tx = db.transaction('kv', 'readwrite');

      tx.oncomplete = function () { callback(); };
      tx.onerror = tx.onabort = function (e) { callback(new Error('Key remove error: ', e.target)); };

      tx.objectStore('kv').delete(key).onerror = function () { tx.abort(); };
    };


    this.set = function (key, value, expire, callback) {
      var tx = db.transaction('kv', 'readwrite');

      tx.oncomplete = function () { callback(); };
      tx.onerror = tx.onabort = function (e) { callback(new Error('Key set error: ', e.target)); };

      tx.objectStore('kv').put({
        key: key,
        value: value,
        expire: expire
      }).onerror = function () { tx.abort(); };
    };


    this.get = function (key, callback) {
      var err, result;
      var tx = db.transaction('kv');

      tx.oncomplete = function () { callback(err, result); };
      tx.onerror = tx.onabort = function (e) { callback(new Error('Key get error: ', e.target)); };

      tx.objectStore('kv').get(key).onsuccess = function(e) {
        if (e.target.result) {
          result = e.target.result.value;
        } else {
          err = new Error('key not found: ' + key);
        }
      };
    };


    this.clear = function (expiredOnly, callback) {
      var keyrange = window.IDBKeyRange; /* ||
                     window.webkitIDBKeyRange || window.msIDBKeyRange;*/
      var tx, store;

      tx = db.transaction('kv', 'readwrite');
      store = tx.objectStore('kv');

      tx.oncomplete = function () { callback(); };
      tx.onerror = tx.onabort = function (e) { callback(new Error('Clear error: ', e.target)); };

      if (expiredOnly) {

        var cursor = store.index('expire').openCursor(keyrange.bound(1, +new Date()));

        cursor.onsuccess = function (e) {
          var _cursor = e.target.result;
          if (_cursor) {
            store.delete(_cursor.primaryKey).onerror = function () { tx.abort(); };
            _cursor.continue();
          }
        };

      } else {
        // Just clear everything
        tx.objectStore('kv').clear().onerror = function () { tx.abort(); };
      }
    };
  };


  Idb.prototype.exists = function() {
    return !!(window.indexedDB /*||
              window.webkitIndexedDB ||
              window.mozIndexedDB ||
              window.msIndexedDB*/);
  };



  /////////////////////////////////////////////////////////////////////////////
  // key/value storage with expiration

  var storeAdapters = {
    'indexeddb': Idb,
    'websql': WebSql,
    'localstorage': DomStorage
  };


  // namespace - db name or similar
  // storesList - array of allowed adapter names to use
  //
  var Storage = function (namespace, storesList) {
    var self = this;

    var db = null;

    // States of db init singletone process
    // 'done' / 'progress' / 'failed' / undefined
    var initState;
    var initStack = [];

    _each(storesList, function(name) {
      // do storage names case insensitive
      name = name.toLowerCase();

      if (!storeAdapters[name]) {
        throw new Error('Wrong storage adapter name: ' + name, storesList);
      }

      if (storeAdapters[name].prototype.exists() && !db) {
        db = new storeAdapters[name](namespace);
        return false; // terminate search on first success
      }
    });

    if (!db) {
      // If no adaprets - don't make error for correct fallback.
      // Just log that we continue work without storing results.
      if (typeof console !== 'undefined' && console.log) {
        console.log('None of requested storages available: ' + storesList);
      }
    }


    this.init = function (callback) {
      if (!db) {
        callback(new Error('No available db'));
        return;
      }

      if (initState === 'done') {
        callback();
        return;
      }

      if (initState === 'progress') {
        initStack.push(callback);
        return;
      }

      initStack.push(callback);
      initState = 'progress';

      db.init(function (err) {
        initState = !err ? 'done' : 'failed';
        _each(initStack, function (cb) {
          cb(err);
        });
        initStack = [];

        // Clear expired. A bit dirty without callback,
        // but we don't care until clear compleete
        if (!err) { self.clear(true); }
      });
    };


    this.set = function (key, value, expire, callback) {
      if (_isFunction(expire)) {
        callback = expire;
        expire = null;
      }
      callback = callback || _nope;
      expire = expire ? +(new Date()) + (expire * 1000) : 0;

      this.init(function(err) {
        if (err) { return callback(err); }
        db.set(key, value, expire, callback);
      });
    };


    this.get = function (key, callback) {
      this.init(function(err) {
        if (err) { return callback(err); }
        db.get(key, callback);
      });
    };


    this.remove = function (key, callback) {
      callback = callback || _nope;
      this.init(function(err) {
        if (err) { return callback(err); }
        db.remove(key, callback);
      });
    };


    this.clear = function (expiredOnly, callback) {
      if (_isFunction(expiredOnly)) {
        callback = expiredOnly;
        expiredOnly = false;
      }
      callback = callback || _nope;

      this.init(function(err) {
        if (err) { return callback(err); }
        db.clear(expiredOnly, callback);
      });
    };
  };


  //////////////////////////////////////////////////////////////////////////////
  // Bag class implementation

  function Bag(options) {
    if (!(this instanceof Bag)) { return new Bag(options); }

    var self = this;

    options = options || {};

    this.prefix       = options.prefix || 'bag';
    this.timeout      = options.timeout || 20;    // 20 seconds
    this.expire       = options.expire || 30 * 24;  // 30 days
    this.isValidItem  = options.isValidItem || null;

    this.stores = _isArray(options.stores) ? options.stores : [ 'indexeddb', 'websql', 'localstorage' ];

    var storage = null;

    this._queue = [];
    this._chained = false;

    this._createStorage = function () {
      if (!storage) { storage = new Storage(self.prefix, self.stores); }
    };

    function getUrl(url, callback) {
      var xhr = new XMLHttpRequest();
      xhr.open('GET', url);
      xhr.onreadystatechange = function() {
        if (xhr.readyState === 4) {
          if (xhr.status === 200) {
            callback(null, {
              content: xhr.responseText,
              type: xhr.getResponseHeader('content-type')
            });
            callback = _nope;
          } else {
            callback(new Error('Can\'t open url ' + url +
               (xhr.status ? xhr.statusText + ' (' + xhr.status + ')' : '')));
            callback = _nope;
          }
        }
      };

      setTimeout(function () {
        if (xhr.readyState < 4) {
          xhr.abort();
          callback(new Error('Timeout'));
          callback = _nope;
        }
      }, self.timeout * 1000);

      xhr.send();
    }

    function createCacheObj(obj, response) {
      var cacheObj = {};

      _each([ 'url', 'key', 'unique' ], function (key) {
        if (obj[key]) { cacheObj[key] = obj[key]; }
      });

      var now = +new Date();
      cacheObj.data = response.content;
      cacheObj.originalType = response.type;
      cacheObj.type = obj.type || response.type;
      cacheObj.stamp = now;

      return cacheObj;
    }

    function saveUrl(obj, callback) {
      getUrl(obj.url_real, function(err, result) {
        if (err) { return callback(err); }

        var delay = (obj.expire || self.expire) * 60 * 60; // in seconds

        var cached = createCacheObj(obj, result);

        self.set(obj.key, cached, delay, function() {
          // Don't check error - have to return data anyway
          _default(obj, cached);
          callback(null, obj);
        });
      });
    }


    function isCacheValid(cached, obj) {
      return !cached ||
        cached.expire - +new Date() < 0  ||
        obj.unique !== cached.unique ||
        obj.url !== cached.url ||
        (self.isValidItem && !self.isValidItem(cached, obj));
    }


    function fetch(obj, callback) {

      if (!obj.url) { return callback(); }
      obj.key = (obj.key || obj.url);

      self.get(obj.key, function(err_cache, cached) {

        // Check error only on forced fetch from cache
        if (err_cache && obj.cached) {
          callback(err_cache);
          return;
        }

        // if can't get object from store, then just load it from web.
        obj.execute = (obj.execute !== false);
        var shouldFetch = !!err_cache || isCacheValid(cached, obj);

        // If don't have to load new date - return one from cache
        if (!obj.live && !shouldFetch) {
          obj.type = obj.type || cached.originalType;
          _default(obj, cached);
          callback(null, obj);
          return;
        }

        // calculate loading url
        obj.url_real = obj.url;
        if (obj.unique) {
          // set parameter to prevent browser cache
          obj.url_real = obj.url + ((obj.url.indexOf('?') > 0) ? '&' : '?') + 'bag-unique=' + obj.unique;
        }

        saveUrl(obj, function(err_load) {
          if (err_cache && err_load) {
            callback(err_load);
            return;
          }

          if (err_load) {
            obj.type = obj.type || cached.originalType;
            _default(obj, cached);
            callback(null, obj);
            return;
          }

          callback(null, obj);
        });
      });
    }

    ////////////////////////////////////////////////////////////////////////////
    // helpers to set absolute sourcemap url

    var sourceMappingRe = /(?:^([ \t]*\/\/[@|#][ \t]+sourceMappingURL=)(.+?)([ \t]*)$)|(?:^([ \t]*\/\*[@#][ \t]+sourceMappingURL=)(.+?)([ \t]*\*\/[ \t])*$)/mg;

    function parse_url(url) {
      var pattern = new RegExp('^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?');
      var matches = url.match(pattern);
      return {
        scheme: matches[2],
        authority: matches[4],
        path: matches[5],
        query: matches[7],
        fragment: matches[9]
      };
    }

    function patchMappingUrl(obj) {
      var refUrl = parse_url(obj.url);
      var done = false;
      var res = obj.data.replace(sourceMappingRe, function(match, p1, p2, p3, p4, p5, p6) {
        if (!match) { return null; }
        done = true;
        // select matched group of params
        if (!p1) { p1 = p4; p2 = p5; p3 = p6; }
        var mapUrl = parse_url(p2);

        var scheme = (mapUrl.scheme ? mapUrl.scheme : refUrl.scheme) || window.location.protocol.slice(0, -1);
        var authority = (mapUrl.authority ? mapUrl.authority : refUrl.authority) || window.location.host;
        var path = mapUrl.path[0] === '/' ? mapUrl.path : refUrl.path.split('/').slice(0, -1).join('/') + '/' + mapUrl.path;
        return p1 + (scheme + '://' + authority + path) + p3;
      });
      return done ? res : '';
    }

    ////////////////////////////////////////////////////////////////////////////

    var handlers = {
      'application/javascript': function injectScript(obj) {
        var script = document.createElement('script'), txt;

        // try to change sourcemap address to absolute
        txt = patchMappingUrl(obj);
        if (!txt) {
          // or add script name for dev tools
          txt = obj.data + '\n//# sourceURL=' + obj.url;
        }

        // Have to use .text, since we support IE8,
        // which won't allow appending to a script
        script.text = txt;
        head.appendChild(script);
        return;
      },

      'text/css': function injectStyle(obj) {
        var style = document.createElement('style'), txt;

        // try to change sourcemap address to absolute
        txt = patchMappingUrl(obj);
        if (!txt) {
          // or add stylesheet script name for dev tools
          txt = obj.data + '\n/*# sourceURL=' + obj.url + ' */';
        }

        if (style.styleSheet) {
          style.styleSheet.cssText = txt; // IE method
        } else {
          style.appendChild(document.createTextNode(txt)); // others
        }

        head.appendChild(style);
        return;
      }
    };


    function execute(obj) {
      if (!obj.type) { return; }

      // Cut off encoding if exists:
      // application/javascript; charset=UTF-8
      var handlerName = obj.type.split(';')[0];

      // Fix outdated mime types if needed, to use single handler
      if ([ 'application/x-javascript', 'text/javascript' ].indexOf(handlerName) >= 0) {
        handlerName = 'application/javascript';
      }

      if (handlers[handlerName]) {
        handlers[handlerName](obj);
      }
      return;
    }

    ////////////////////////////////////////////////////////////////////////////

    //
    // Public methods
    //

    this.require = function(resources, callback) {
      var queue = self._queue;

      if (_isFunction(resources)) {
        callback = resources;
        resources = null;
      }

      if (resources) {
        var res = _isArray(resources) ? resources : [ resources ];

        // convert string urls to structures
        // and push to queue
        _each(res, function(r, i) {
          if (_isString(r)) { res[i] = { url: r }; }
          queue.push(res[i]);
        });
      }

      self._createStorage();

      if (!callback) {
        self._chained = true;
        return self;
      }

      _asyncEach(queue, fetch, function(err) {
        if (err) {
          // cleanup
          self._chained = false;
          self._queue = [];

          callback(err);
          return;
        }

        _each(queue, function(obj) {
          if (obj.execute) {
            execute(obj);
          }
        });

        // return content only, if one need fuul info -
        // check input object, that will be extended.
        var replies = [];
        _each(queue, function(r) { replies.push(r.data); });

        var result = (_isArray(resources) || self._chained) ? replies : replies[0];

        // cleanup
        self._chained = false;
        self._queue = [];

        callback(null, result);
      });
    };


    // Create proxy methods (init store then subcall)
    _each([ 'remove', 'get', 'set', 'clear' ], function(method) {
      self[method] = function () {
        self._createStorage();
        storage[method].apply(storage, arguments);
      };
    });


    this.addHandler = function (types, handler) {
      types = _isArray(types) ? types : [ types ];
      _each(types, function (type) { handlers[type] = handler; });
    };


    this.removeHandler = function (types) {
      self.addHandler(types/*, undefined*/);
    };
  }

  return Bag;

}));
